File: 백준 18870 좌표 압축 -정렬-/Main.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class Main {
    private static Map<Integer, Integer> map = new HashMap<>();
    private static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int[] nums = new int[N];
        int[] sortNums = new int[N];
        for (int i = 0; i < N; i++) {
            int num = sc.nextInt();
            nums[i] = num;
            sortNums[i] = num;
        }
        //오름차순 정렬
        Arrays.sort(sortNums);
        int cnt = 0;
        for (int i = 0; i < N; i++) {
            if (map.get(sortNums[i]) == null) { //아직 압축 안된 숫자인 경우
                map.put(sortNums[i], cnt++);
            }
        }

        for (int i = 0; i < N; i++) {
            sb.append(map.get(nums[i])).append(" ");
        }
        System.out.println(sb.toString());
    }
}

File: 백준 18870 좌표 압축 -정렬-/Main.kt
import java.util.*


private val map: MutableMap<Int, Int> = HashMap()
private val sb = StringBuilder()

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    val N = sc.nextInt()
    val nums = IntArray(N)
    val sortNums = IntArray(N)
    for (i in 0 until N) {
        val num = sc.nextInt()
        nums[i] = num
        sortNums[i] = num
    }
    //오름차순 정렬
    Arrays.sort(sortNums)
    var cnt = 0
    for (i in 0 until N) {
        if (map[sortNums[i]] == null) { //아직 압축 안된 숫자인 경우
            map[sortNums[i]] = cnt++
        }
    }
    for (i in 0 until N) {
        sb.append(map[nums[i]]).append(" ")
    }
    println(sb.toString())
}
